src/util/Cache.h

```diff
@@ -0,0 +1,366 @@
+                         AccessUpdater accessUpdater,
+                         ScoreCalculator scoreCalculator)
+      : _capacity(capacity),
+        _data(scoreComparator),
+        _accessUpdater(accessUpdater),
+        _scoreCalculator(scoreCalculator) {}
+
+  // tryEmplace allows for race-free adding of items to the cache. Iff no item
+  // in the cache is associated with the key a new item is created and
+  // a writeable reference is returned as the first element of the result pair
+  // and it is the callers responsibility to synchronize access when writing to
+  // it. If there already was an item associated with the key the first element
+  // is shared_ptr<Value>(nullptr) and the second provides read-only access.
+  //
+  // The second element of the pair always returns a valid read-only reference,
+  // either to the just created element or an existing element.
+  //
+  // This needs to happen in a single operation because otherwise we may find
+  // an item in the cash now but it may already be deleted when trying to
+  // retrieve  it in another operation.
+  //
+  // If the optional parameter pin is true the emplaced element will
+  // never be thrown out of the cache unless removed explicitly.
+  template <class... Args>
+  TryEmplaceResult tryEmplace(const Key& key, Args&&... args) {
+    std::lock_guard lock(_lock);
+
+    if (const auto pinnedIt = _pinnedMap.find(key);
+        pinnedIt != _pinnedMap.end()) {
+      return TryEmplaceResult(shared_ptr<Value>(nullptr), pinnedIt->second);
+    }
+
+    if (const auto mapIt = _accessMap.find(key); mapIt != _accessMap.end()) {
+      auto& handle = mapIt->second;
+      // Move element to the front as it is now least recently used
+      // The handle changes. In the current implementation, only irrelevant
+      // parts of the handles should change, but we still do the update
+      // TODO<joka921> : measure if this is performance critical.
+      _data.updateKey(_accessUpdater(handle.score(), handle.value()), &handle);
+      return TryEmplaceResult(EmplacedValue(nullptr),
+                              _accessMap[key].value().value());
+    }
+
+    // Insert without taking mutex recursively
+    EmplacedValue emplaced = make_shared<Value>(std::forward<Args>(args)...);
+
+    _accessMap[key] =
+        _data.insert(_scoreCalculator(*emplaced), Entry(key, emplaced));
+    if (_data.size() > _capacity) {
+      // Remove the last element.
+      // Since we are using shared_ptr this does not free the underlying
+      // memory if it is still accessible through a previously returned
+      // shared_ptr
+
+      _accessMap.erase(_data.pop().value().key());
+    }
+    assert(_data.size() <= _capacity);
+    return TryEmplaceResult(emplaced, emplaced);
+  }
+
+  // tryEmplacePinned acts like tryEmplace but if the pin paramater is true
+  // also marks the emplaced element so that it can only be removed explicitly
```

Comment:
This was actually my mistake in the original code but let's fix it now, there is no "pinned parameter" that was the original idea but varargs made it impossible